x-pack/plugins/monitoring/server/plugin.ts

```diff
@@ -0,0 +1,324 @@
+      id: 'monitoring',
+      name: i18n.translate('xpack.monitoring.featureRegistry.monitoringFeatureName', {
+        defaultMessage: 'Stack Monitoring',
+      }),
+      icon: 'monitoringApp',
+      navLinkId: 'monitoring',
+      app: ['monitoring', 'kibana'],
+      catalogue: ['monitoring'],
+      privileges: {},
+      reserved: {
+        privilege: {
+          savedObject: {
+            all: [],
+            read: [],
+          },
+          ui: [],
+        },
+        description: i18n.translate('xpack.monitoring.feature.reserved.description', {
+          defaultMessage: 'To grant users access, you should also assign the monitoring_user role.',
+        }),
+      },
+    });
+  }
+
+  async setupLegacy(
+    legacyApi: LegacyAPI,
+    core: CoreSetup,
+    monitoringCore: MonitoringCore,
+    config: MonitoringConfig,
+    plugins: PluginsSetup,
+    cluster: ICustomClusterClient
+  ) {
+    const legacyConfigWrapper = monitoringCore.config
+      ? monitoringCore.config()
+      : { get: () => null };
+    const log = this.getLogger(LOGGING_TAG, KIBANA_MONITORING_LOGGING_TAG);
+
+    // Initialize telemetry
+    registerMonitoringCollection(cluster, legacyApi.telemetryCollectionManager);
+    // Register collector objects for stats to show up in the APIs
+    registerCollectors(plugins.usageCollection, {
+      elasticsearchPlugin: legacyApi.elasticsearch,
+      interval: legacyApi.opsInterval,
+      log,
+      config: legacyConfigWrapper,
+      getOSInfo: legacyApi.getOSInfo,
+      hapiServer: legacyApi.hapiServer,
+      metrics$: core.metrics.getOpsMetrics$(),
+    });
+
+    // If collection is enabled, create the bulk uploader
+    const kibanaCollectionEnabled = config.kibana.collection.enabled;
+    if (kibanaCollectionEnabled) {
+      // Start kibana internal collection
+      const serverInfo = core.http.getServerInfo();
+      const legacyConfig = await this.initializerContext.config.legacy.globalConfig$
+        .pipe(first())
+        .toPromise();
+      const bulkUploader = initBulkUploader({
+        elasticsearch: core.elasticsearch,
+        config,
+        log,
+        kibanaStats: {
+          uuid: core.uuid.getInstanceUuid(),
+          name: serverInfo?.name,
+          index: get(legacyConfig, 'kibana.index'),
+          host: serverInfo?.host,
+          port: serverInfo?.port?.toString(),
+          version: this.initializerContext.env.packageInfo.version,
+          status: legacyApi.getServerStatus(),
+        },
+      });
+
+      // Do not use `this.licenseService` as that looks at the monitoring cluster
+      // whereas we want to check the production cluster here
+      plugins.licensing.license$.subscribe((license: any) => {
+        // use updated xpack license info to start/stop bulk upload
+        const mainMonitoring = license.getFeature('monitoring');
+        const monitoringBulkEnabled =
+          mainMonitoring && mainMonitoring.isAvailable && mainMonitoring.isEnabled;
```

Comment:
can be moved to licensing service?

Reply:
Possibly, but the licensing service (in this plugin) deals with the monitoring cluster, whereas this code needs to look at the production cluster so I figured it made sense to keep them different